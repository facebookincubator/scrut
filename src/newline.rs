/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! This module provides convenient handlers for handling common newline relating
//! operations, like assuring that strings do end (or do not end) in a newline.
//! Also, since Scrut deals with raw byte output of scripts, that may not even
//! be unicode, the whole shabang is implemented for byte arrays as well.
//!
//! Regarding CRLF and LF line endings:
//! Scrut internally only works with LF endings and considers CRLF only at
//! certain boundaries, which are:
//!
//! - Content read from test files: CRLF is converted into LF
//! - Output generated from test execution: CRLF is converter into LF
//!
//! In line with the Rust policy, Scrut does _NOT_ output CRLF at any time.
use std::borrow::Cow;

/// Extends byte slices to provide easy interface dealing with newline characters
pub(crate) trait BytesNewline {
    /// Return the byte slice without tailing newline character(s)
    fn trim_newlines(&self) -> &[u8];

    /// Returns byte slice that ends in newline character(s)
    fn assure_newline(&self) -> Cow<'_, [u8]>;

    /// Return bool whether ends in new line
    fn ends_in_newline(&self) -> bool;
}

impl BytesNewline for &[u8] {
    fn trim_newlines(&self) -> &[u8] {
        let mut right_index = self.len();
        while ends_in_newline(&self[0..right_index]) {
            right_index -= 1;
        }
        &self[0..right_index]
    }

    fn assure_newline(&self) -> Cow<'_, [u8]> {
        if ends_in_newline(self) {
            return (*self).into();
        }
        let mut updated = self.to_vec();
        updated.push(b'\n');
        updated.into()
    }

    fn ends_in_newline(&self) -> bool {
        ends_in_newline(self)
    }
}

fn ends_in_newline(data: &[u8]) -> bool {
    let l = data.len();
    l > 0 && data[l - 1] == b'\n'
}

/// Extend strings with interface to assure
pub(crate) trait StringNewline {
    fn trim_newlines(&self) -> Cow<'_, str>;
    fn assure_newline(&self) -> Cow<'_, str>;
}

impl StringNewline for &str {
    fn assure_newline(&self) -> Cow<'_, str> {
        assure_newline(self)
    }

    fn trim_newlines(&self) -> Cow<'_, str> {
        trim_newlines(self)
    }
}

impl StringNewline for String {
    fn assure_newline(&self) -> Cow<'_, str> {
        assure_newline(self)
    }

    fn trim_newlines(&self) -> Cow<'_, str> {
        trim_newlines(self)
    }
}

fn assure_newline(line: &str) -> Cow<'_, str> {
    if line.ends_with('\n') {
        line.into()
    } else {
        (line.to_string() + "\n").into()
    }
}

fn trim_newlines(line: &'_ str) -> Cow<'_, str> {
    if let Some(stripped) = line.strip_suffix('\n') {
        trim_newlines(stripped)
    } else {
        line.into()
    }
}

/// A helper trait that provides a convenient way to split byte arrays into lines
pub(crate) trait SplitLinesByNewline {
    /// Splits byte array by LF while keeping the ending LF for each item
    /// of the returned vector
    fn split_at_newline(&self) -> Vec<&[u8]>;
}

impl SplitLinesByNewline for &[u8] {
    fn split_at_newline(&self) -> Vec<&[u8]> {
        split_at_newline(self)
    }
}

fn split_at_newline(text: &[u8]) -> Vec<&[u8]> {
    let mut lines = vec![];
    let mut start = 0;

    for (index, byte) in text.iter().enumerate() {
        if *byte == b'\n' {
            lines.push(&text[start..index + 1]);
            start = index + 1;
        }
    }
    if start < text.len() {
        lines.push(&text[start..])
    }

    lines
}

const CRLF: &[u8] = b"\r\n";

/// Replaces all CRLF with LF
pub fn replace_crlf<'a>(bytes: &'a [u8]) -> Cow<'a, [u8]> {
    if let Some(index) = bytes.windows(2).position(|window| window == CRLF) {
        [
            Cow::from(&bytes[0..index]),
            replace_crlf(&bytes[index + 1..]),
        ]
        .concat()
        .into()
    } else {
        bytes.into()
    }
}

/// Like the [`format`] with an added new line character
#[macro_export]
macro_rules! formatln {
    ($arg:expr) => {{
        format!("{}\n", $arg)
    }};
    ($arg:tt, $($args:tt)*) => {{
        format!("{}\n", format!($arg, $($args)*))
    }};
}

/// Like the [`format`] with an added new line character and in raw bytes
#[macro_export]
macro_rules! bformatln {
    ($arg:expr) => {{
        format!("{}\n", $arg).as_bytes().to_vec()
    }};
    ($arg:tt, $($args:tt)*) => {{
        format!("{}\n", format!($arg, $($args)*)).as_bytes().to_vec()
    }};
}

#[macro_export]
macro_rules! blines {
    ($arg:tt) => {{
        format!("{}\n", $arg).as_bytes().to_vec()
    }};
    ($arg:tt, $($args:tt)*) => {{
        ([$arg, $($args)*].join("\n") + "\n").as_bytes().to_vec()
    }};
}

#[cfg(test)]
mod tests {
    use super::BytesNewline;
    use super::SplitLinesByNewline;
    use super::StringNewline;
    use super::replace_crlf;
    use crate::newline::assure_newline;

    #[test]
    fn test_formatln() {
        assert_eq!(formatln!("something"), "something\n");
        assert_eq!(
            formatln!("something {}", "bla"),
            "something bla\n".to_string()
        );
        assert_eq!(
            formatln!("something {} {}", "bla", "foo"),
            "something bla foo\n".to_string(),
        );
    }

    #[test]
    fn test_bformatln() {
        assert_eq!(bformatln!("something"), "something\n".as_bytes());
        assert_eq!(
            bformatln!("something {}", "bla"),
            "something bla\n".as_bytes(),
        );
        assert_eq!(
            bformatln!("something {} {}", "bla", "foo"),
            "something bla foo\n".as_bytes()
        );
    }

    #[test]
    fn test_blines() {
        assert_eq!(blines!("something"), "something\n".as_bytes(),);
        assert_eq!(blines!("something", "bla"), "something\nbla\n".as_bytes());
        assert_eq!(
            blines!("something", "bla", "foo"),
            "something\nbla\nfoo\n".as_bytes()
        );
    }

    #[test]
    fn test_trim_newlines_bytes() {
        let tests = vec![("foo", "foo"), ("foo\n", "foo"), ("foo\n\n\n", "foo")];

        for (from, expect) in tests {
            let fromb = from.as_bytes();
            let to = fromb.trim_newlines();
            assert_eq!(expect.as_bytes(), to, "from `{:?}`", fromb);
        }
    }

    #[test]
    fn test_assure_newline_bytes() {
        let tests = vec![
            ("foo", "foo\n"),
            ("foo\n", "foo\n"),
            ("foo\n\n\n", "foo\n\n\n"),
        ];

        for (from, expect) in tests {
            let fromb = from.as_bytes();
            let to = fromb.assure_newline().to_vec();
            assert_eq!(expect.as_bytes().to_vec(), to, "from `{:?}`", fromb);
        }
    }

    #[test]
    fn test_trim_newlines_string() {
        let tests = vec![("foo", "foo"), ("foo\n", "foo"), ("foo\n\n\n", "foo")];
        for (from, expect) in tests {
            let result = from.trim_newlines();
            assert_eq!(expect, result, "from {}", from);
        }
    }

    #[test]
    fn test_assure_newline_string() {
        let tests = vec![
            ("foo", "foo\n"),
            ("foo\n", "foo\n"),
            ("foo\n\n\n", "foo\n\n\n"),
        ];
        for (from, expect) in tests {
            let result = assure_newline(from);
            assert_eq!(expect, &result, "from {}", from);
        }
    }

    #[test]
    fn test_split_at_newline() {
        let tests = vec![
            ("l1", vec![b"l1" as &[u8]]),
            ("l1\n", vec![b"l1\n"]),
            ("l1\nl2", vec![b"l1\n", b"l2"]),
            ("l1\nl2\nl3\n", vec![b"l1\n", b"l2\n", b"l3\n"]),
        ];
        for (from, expect) in tests {
            let to = from.as_bytes();
            let to = to.split_at_newline();
            assert_eq!(expect, to, "from `{}`", from)
        }
    }

    #[test]
    fn test_replace_crlf() {
        let tests = vec![
            ("l1", "l1"),
            ("l1\nl2", "l1\nl2"),
            ("l1\r\nl2", "l1\nl2"),
            ("l1\r\nl2\r\n", "l1\nl2\n"),
            ("\r\nl1\r\nl2\r\n", "\nl1\nl2\n"),
        ];

        for (from, expect) in tests {
            let to = replace_crlf(from.as_bytes());
            assert_eq!(expect.as_bytes().to_vec(), to.to_vec(),);
        }
    }
}
