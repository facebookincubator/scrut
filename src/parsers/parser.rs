use std::fmt::Display;

use anyhow::Result;
use clap::ValueEnum;

use crate::testcase::TestCase;

/// A Parser extracts one or more [`crate::testcase::TestCase`]s from a provided text
pub trait Parser {
    /// Returns all testcases found in the provided text
    fn parse(&self, tests: &str) -> Result<Vec<TestCase>>;
}

#[derive(Debug, Clone, Copy, PartialEq, ValueEnum)]
pub enum ParserType {
    Markdown,
    Cram,
}

impl ParserType {
    pub fn file_extension(&self) -> &'static str {
        match self {
            Self::Cram => "t",
            Self::Markdown => "md",
        }
    }
}

impl Display for ParserType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Cram => "cram",
                Self::Markdown => "markdown",
            }
        )
    }
}

impl TryFrom<String> for ParserType {
    type Error = String;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        match &value as &str {
            "markdown" | "md" => Ok(Self::Markdown),
            "cram" => Ok(Self::Cram),
            _ => Err(format!("Unsupported parser format `{value}`")),
        }
    }
}
