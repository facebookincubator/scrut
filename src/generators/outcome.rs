use anyhow::bail;
use anyhow::Result;

use crate::diff::DiffLine;
use crate::formatln;
use crate::lossy_string;
use crate::newline::BytesNewline;
use crate::newline::SplitLinesByNewline;
use crate::newline::StringNewline;
use crate::outcome::Outcome;
use crate::output::ExitStatus;
use crate::testcase::TestCaseError;

pub(super) trait OutcomeTestGenerator {
    fn generate_testcase(&self) -> Result<String>;
}

impl Outcome {
    fn generate_testcase_expression(&self) -> String {
        // prepend by command
        let expression_lines = self.testcase.shell_expression.as_bytes();
        let expression_lines = expression_lines.split_at_newline();
        let mut generated = format!("$ {}", lossy_string!(&expression_lines[0].assure_newline()));
        expression_lines.iter().skip(1).for_each(|line| {
            generated.push_str(&format!(
                "> {}",
                lossy_string!(&(&line[..]).assure_newline())
            ))
        });
        generated
    }

    fn generate_testcase_exit_code(&self) -> Option<String> {
        match &self.output.exit_code {
            ExitStatus::Code(code) if *code != 0 => Some(formatln!("[{}]", code)),
            _ => None,
        }
    }
}

impl OutcomeTestGenerator for Outcome {
    fn generate_testcase(&self) -> Result<String> {
        match &self.result {
            Ok(_) => {
                let mut generated = self.generate_testcase_expression();
                self.testcase.expectations.iter().for_each(|expectation| {
                    generated.push_str(&expectation.original_string().assure_newline())
                });
                if let Some(exit_code) = self.generate_testcase_exit_code() {
                    generated.push_str(&exit_code)
                }
                Ok(generated)
            }
            Err(err) => match err {
                TestCaseError::MalformedOutput(diff) => {
                    let mut generated = self.generate_testcase_expression();

                    // output the actual recorded output lines
                    for diff_line in diff.lines.iter() {
                        match diff_line {
                            DiffLine::MatchedExpectation {
                                index: _,
                                expectation,
                                lines: _,
                            } => {
                                generated.push_str(&expectation.original_string().assure_newline())
                            }
                            DiffLine::UnexpectedLines { lines } => {
                                for (_, line) in lines {
                                    let suffix = if line.ends_with(b"\n") {
                                        ""
                                    } else {
                                        " (no-eol)"
                                    };
                                    let line = formatln!(
                                        "{}{}",
                                        self.escaping
                                            .escaped_expectation((&line[..]).trim_newlines()),
                                        suffix
                                    );
                                    generated.push_str(&line)
                                }
                            }
                            _ => continue,
                        }
                    }
                    if let Some(exit_code) = self.generate_testcase_exit_code() {
                        generated.push_str(&exit_code)
                    }
                    Ok(generated)
                }
                TestCaseError::InvalidExitCode {
                    actual,
                    expected: _,
                } => {
                    let mut generated = self.generate_testcase_expression();
                    let mut output = self.output.stdout.to_output_string(None, &self.escaping);
                    if !output.is_empty() && !output.ends_with('\n') {
                        output.push_str(" (no-eol)\n")
                    }
                    generated.push_str(&output);
                    generated.push_str(&formatln!("[{}]", *actual));
                    Ok(generated)
                }
                TestCaseError::InternalError(err) => {
                    bail!("cannot generate testcase from internal error: {}", err)
                }
                TestCaseError::Skipped => {
                    bail!("cannot generate skipped testcase")
                }
            },
        }
    }
}
