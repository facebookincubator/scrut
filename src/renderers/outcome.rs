/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use anyhow::Result;
use colored::ColoredString;
use colored::Colorize;

use crate::formatln;
use crate::outcome::Outcome;

const MAX_LINE_LENGTH: usize = 80;

pub(super) trait OutcomeHeader {
    fn render_header(&self) -> Result<String>;
}

impl OutcomeHeader for Outcome {
    fn render_header(&self) -> Result<String> {
        let mut headers = vec![];
        if let Some(ref location) = self.location {
            headers.push(header_to_title(
                "@",
                &format!("{}:{}", location, self.testcase.line_number),
                |s| s.bright_blue(),
            ));
        } else {
            headers.push(header_to_title(
                "@",
                &format!("Line {}", self.testcase.line_number),
                |s| s.bright_blue(),
            ));
        }
        if !self.testcase.title.is_empty() {
            headers.push(header_to_title("#", &self.testcase.title, |s| {
                s.bright_cyan()
            }));
        }
        headers.push(header_to_title("$", &self.testcase.shell_expression, |s| {
            s.bold().bright_yellow()
        }));

        let divider_outer = &formatln!("// {}", "=".repeat(MAX_LINE_LENGTH - 3))
            .bright_black()
            .to_string();
        let divider_inner = &formatln!("// {}", "-".repeat(MAX_LINE_LENGTH - 3))
            .bright_black()
            .to_string();

        let mut output = String::new();
        output.push_str(divider_outer);
        output.push_str(&headers.join(divider_inner));
        output.push_str(divider_outer);
        output.push('\n');

        Ok(output)
    }
}

fn header_to_title(first_prefix: &str, text: &str, color: fn(&str) -> ColoredString) -> String {
    let mut title = String::new();
    let prefix = "//".bright_black();
    for (index, line) in text.split('\n').enumerate() {
        title.push_str(&formatln!(
            "{} {} {}",
            prefix,
            (if index == 0 { first_prefix } else { " " }).bold(),
            color(line)
        ))
    }
    title
}
