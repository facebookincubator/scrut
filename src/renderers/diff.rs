use std::cmp::Ordering;
use std::fmt::Display;

use anyhow::bail;
use anyhow::Result;

use super::renderer::ErrorRenderer;
use super::renderer::Renderer;
use crate::diff::Diff;
use crate::diff::DiffLine;
use crate::formatln;
use crate::newline::BytesNewline;
use crate::outcome::Outcome;
use crate::parsers::parser::ParserType;

/// Renderer that uses the traditional Diff render format
/// See: https://en.wikipedia.org/wiki/Diff
pub struct DiffRenderer {}

impl DiffRenderer {
    pub fn new() -> Self {
        Self {}
    }
}

impl Default for DiffRenderer {
    fn default() -> Self {
        Self::new()
    }
}

impl Renderer for DiffRenderer {
    fn render(&self, outcomes: &[&Outcome]) -> Result<String> {
        let mut output = String::new();
        let count_locations = outcomes
            .iter()
            .filter(|outcome| outcome.location.is_some())
            .count();
        let mut outcomes = outcomes.to_owned().to_vec();
        if count_locations > 0 {
            if count_locations != outcomes.len() {
                bail!("cannot render diff with some outcomes providing locations, but not all")
            }
            outcomes.sort_by(|a, b| {
                let result = a.location.cmp(&b.location);
                if result == Ordering::Equal {
                    a.testcase.line_number.cmp(&b.testcase.line_number)
                } else {
                    result
                }
            })
        }
        let mut last_location = None;
        for outcome in outcomes {
            match &outcome.result {
                Ok(_) => continue,
                Err(err) => {
                    if outcome.location != last_location {
                        if let Some(ref location) = outcome.location {
                            if last_location.is_some() {
                                output.push('\n');
                            }
                            last_location = outcome.location.to_owned();
                            output.push_str(&formatln!("--- {}", location));
                            output.push_str(&formatln!("+++ {}.new", location));
                        }
                    }
                    let rendered_error = self.render_error(err, outcome)?;
                    output.push_str(&rendered_error);
                }
            }
        }
        Ok(output)
    }
}

impl ErrorRenderer for DiffRenderer {
    /// Render result contains a change for the last line of the output, which
    /// would contain the exit code `[<exit-code>]`.
    /// The rest of the output is *not* attended, even if it is wrong.
    fn render_invalid_exit_code(
        &self,
        outcome: &Outcome,
        actual: i32,
        _expected: i32,
    ) -> Result<String> {
        let line_number = outcome.testcase.line_number
            + outcome.testcase.shell_expression_lines()
            + outcome.testcase.expectations_lines();
        let prefix = line_prefix(outcome);
        let title = join_multiline(&outcome.testcase.title, " * ");
        let mut output = String::new();
        output.push_str(
            &DiffHeader {
                old_start: line_number,
                old_length: outcome.testcase.exit_code.map_or(0, |_| 1),
                new_start: line_number,
                new_length: 1,
                kind: DiffHeaderKind::InvalidExitCode,
                title: &title,
            }
            .to_string(),
        );

        if let Some(exit_code) = outcome.testcase.exit_code {
            output.push_str(&format!("-{prefix}[{exit_code}]\n"));
        }
        output.push_str(&format!("+{prefix}[{actual}]\n"));
        Ok(output)
    }

    /// Renders the internal error, in a way that is NOT compatible with
    /// the unified diff syntax and will be rejected by `patch`. This is
    /// intentional - the error must be handled by a user.
    fn render_delegated_error(&self, outcome: &Outcome, err: &anyhow::Error) -> Result<String> {
        let title = join_multiline(&outcome.testcase.title, " * ");
        let mut output = String::new();
        output.push_str("# ---- INTERNAL ERROR ----\n");
        if let Some(ref location) = outcome.location {
            output.push_str(&format!("# PATH:  {location}\n"));
        }
        output.push_str(&format!("# TITLE: {title}\n"));
        let output_err = err
            .to_string()
            .lines()
            .map(|line| format!("# ERROR: {line}\n"))
            .collect::<Vec<_>>()
            .join("");
        output.push_str(&output_err);
        output.push_str("# ---- INTERNAL ERROR ----\n");
        Ok(output)
    }

    fn render_malformed_output(&self, outcome: &Outcome, diff: &Diff) -> Result<String> {
        UnifiedDiff::default().render(outcome, diff)
    }

    fn render_skipped(&self, _outcome: &Outcome) -> Result<String> {
        Ok("".into())
    }
}

#[derive(Default)]
struct UnifiedDiff {
    unmatched_start: Option<usize>,
    unmatched_lines: Vec<String>,
    unexpected_start: Option<usize>,
    unexpected_lines: Vec<(usize, String)>,
}

impl UnifiedDiff {
    fn flush(&mut self) {
        self.unmatched_start = None;
        self.unexpected_start = None;
        self.unmatched_lines = vec![];
        self.unexpected_lines = vec![];
    }

    fn render(&mut self, outcome: &Outcome, diff: &Diff) -> Result<String> {
        let line_number = outcome.testcase.line_number + outcome.testcase.shell_expression_lines();
        let title = join_multiline(&outcome.testcase.title, " * ");
        let prefix = line_prefix(outcome);
        let mut output = String::new();

        macro_rules! add_diff_hunk {
            () => {
                if self.unmatched_start.is_some() || self.unexpected_start.is_some() {
                    let (unmatched_start, unexpected_start) =
                        if let Some(unmatched_start) = self.unmatched_start {
                            (
                                unmatched_start,
                                self.unexpected_start.unwrap_or(unmatched_start),
                            )
                        } else {
                            let unexpected_start = self.unexpected_start.unwrap();
                            (unexpected_start, unexpected_start)
                        };
                    output.push_str(
                        &DiffHeader {
                            old_start: unmatched_start + line_number,
                            old_length: self.unmatched_lines.len(),
                            new_start: unexpected_start + line_number,
                            new_length: self.unexpected_lines.len(),
                            kind: DiffHeaderKind::MalformedOutput,
                            title: &title,
                        }
                        .to_string(),
                    );
                    self.unmatched_lines
                        .iter()
                        .for_each(|line| output.push_str(&format!("-{prefix}{line}\n")));
                    self.unexpected_lines
                        .iter()
                        .for_each(|line| output.push_str(&format!("+{}{}\n", prefix, &line.1)));
                    self.flush();
                }
            };
        }

        let mut expectation_index = 0;
        for line in &diff.lines {
            match line {
                DiffLine::MatchedExpectation {
                    index,
                    expectation: _,
                    lines: _,
                } => {
                    expectation_index = *index;
                    add_diff_hunk!();
                }
                DiffLine::UnmatchedExpectation { index, expectation } => {
                    expectation_index = *index;
                    if self.unmatched_start.is_none() {
                        self.unmatched_start = Some(*index);
                    }
                    self.unmatched_lines.push(expectation.original_string())
                }
                DiffLine::UnexpectedLines { lines } => {
                    if self.unexpected_start.is_none() {
                        self.unexpected_start = Some(expectation_index)
                    }
                    self.unexpected_lines.extend(
                        lines
                            .iter()
                            .map(|(i, l)| {
                                Ok((
                                    *i,
                                    String::from_utf8((l as &[u8]).trim_newlines().to_vec())?,
                                ))
                            })
                            .collect::<Result<Vec<_>>>()?,
                    );
                    if self.unmatched_start.is_some() {
                        add_diff_hunk!();
                    }
                }
            }
        }
        add_diff_hunk!();

        Ok(output)
    }
}

enum DiffHeaderKind {
    InvalidExitCode,
    MalformedOutput,
}

impl Display for DiffHeaderKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                DiffHeaderKind::InvalidExitCode => "invalid exit code",
                DiffHeaderKind::MalformedOutput => "malformed output",
            }
        )
    }
}

struct DiffHeader<'a> {
    old_start: usize,
    old_length: usize,
    new_start: usize,
    new_length: usize,
    title: &'a str,
    kind: DiffHeaderKind,
}

impl<'a> Display for DiffHeader<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(
            f,
            "@@ -{}{} +{}{} @@ {}: {}",
            self.old_start,
            length_suffix(self.old_length),
            self.new_start,
            length_suffix(self.new_length),
            self.kind,
            self.title,
        )
    }
}

fn length_suffix(len: usize) -> String {
    if len > 1 || len == 0 {
        format!(",{len}")
    } else {
        "".into()
    }
}

fn join_multiline(text: &str, sep: &str) -> String {
    text.lines().collect::<Vec<_>>().join(sep)
}

fn line_prefix(outcome: &Outcome) -> &'static str {
    match outcome.format {
        ParserType::Markdown => "",
        ParserType::Cram => "  ",
    }
}

#[cfg(test)]
mod tests {
    use super::DiffRenderer;
    use crate::diff::Diff;
    use crate::diff::DiffLine;
    use crate::escaping::Escaper;
    use crate::outcome::Outcome;
    use crate::parsers::parser::ParserType;
    use crate::renderers::renderer::Renderer;
    use crate::test_expectation;
    use crate::testcase::TestCase;
    use crate::testcase::TestCaseError;

    #[test]
    fn test_render_success() {
        let renderer = DiffRenderer::new();
        let rendered = renderer
            .render(&[&Outcome {
                output: ("the stdout", "the stderr").into(),
                testcase: TestCase {
                    title: "the title".to_string(),
                    shell_expression: "the command".to_string(),
                    expectations: vec![],
                    exit_code: None,
                    line_number: 234,
                    ..Default::default()
                },
                location: Some("the location".to_string()),
                result: Ok(()),
                escaping: Escaper::default(),
                format: ParserType::Markdown,
            }])
            .expect("render succeeds");
        assert_eq!("", &rendered, "success results are not rendered");
    }

    #[test]
    fn test_internal_error() {
        let renderer = DiffRenderer::new();
        let rendered = renderer
            .render(&[&Outcome {
                output: ("the stdout", "the stderr", Some(222)).into(),
                testcase: TestCase {
                    title: "the title".into(),
                    shell_expression: "the command".into(),
                    expectations: vec![],
                    exit_code: Some(111),
                    line_number: 234,
                    ..Default::default()
                },
                location: Some("the location".into()),
                result: Err(TestCaseError::InternalError(anyhow::Error::msg(
                    "bad thing",
                ))),
                escaping: Escaper::default(),
                format: ParserType::Markdown,
            }])
            .expect("render succeeds");
        insta::assert_snapshot!(rendered);
    }

    #[test]
    fn test_invalid_exit_code() {
        let renderer = DiffRenderer::new();

        [ParserType::Cram, ParserType::Markdown]
            .iter()
            .for_each(|parser_type| {
                let rendered = renderer
                    .render(&[&Outcome {
                        output: ("the stdout\n", "the stderr\n", Some(222)).into(),
                        testcase: TestCase {
                            title: "the title".into(),
                            shell_expression: "the command".into(),
                            expectations: vec![test_expectation!("the stdout")],
                            exit_code: Some(111),
                            line_number: 234,
                            ..Default::default()
                        },
                        location: Some("the location".into()),
                        result: Err(TestCaseError::InvalidExitCode {
                            actual: 222,
                            expected: 111,
                        }),
                        escaping: Escaper::default(),
                        format: *parser_type,
                    }])
                    .expect("render succeeds");
                insta::assert_snapshot!(format!("invalid_exit_code_{parser_type}"), rendered);
            });
    }

    #[test]
    fn test_malformed_output() {
        let renderer = DiffRenderer::new();
        let testcase = TestCase {
            title: "the title".into(),
            shell_expression: "the command".into(),
            expectations: vec![
                test_expectation!("expected line 1"),
                test_expectation!("expected line 2"),
                test_expectation!("expected line 3"),
            ],
            exit_code: None,
            line_number: 234,
            ..Default::default()
        };

        let tests = vec![
            (
                "missing",
                Diff::new(vec![DiffLine::UnmatchedExpectation {
                    index: 1,
                    expectation: testcase.expectations[1].clone(),
                }]),
            ),
            (
                "unexpected",
                Diff::new(vec![DiffLine::UnexpectedLines {
                    lines: vec![(2, "something else".as_bytes().to_vec())],
                }]),
            ),
            (
                "mismatch",
                Diff::new(vec![
                    DiffLine::UnmatchedExpectation {
                        index: 1,
                        expectation: testcase.expectations[1].clone(),
                    },
                    DiffLine::UnmatchedExpectation {
                        index: 2,
                        expectation: testcase.expectations[2].clone(),
                    },
                    DiffLine::UnexpectedLines {
                        lines: vec![
                            (2, "something line 1".as_bytes().to_vec()),
                            (3, "something line 2".as_bytes().to_vec()),
                        ],
                    },
                ]),
            ),
        ];

        [ParserType::Cram, ParserType::Markdown]
            .iter()
            .for_each(|parser_type| {
                tests.iter().for_each(|(name, diff)| {
                    let rendered = renderer
                        .render(&[&Outcome {
                            output: (
                                "expected line 1\nexpected line FAIL\nexpected line 3\n",
                                "the stderr",
                            )
                                .into(),
                            testcase: testcase.clone(),
                            location: Some("the location".into()),
                            result: Err(TestCaseError::MalformedOutput(diff.to_owned())),
                            escaping: Escaper::default(),
                            format: *parser_type,
                        }])
                        .expect("render succeeds");
                    insta::assert_snapshot!(
                        format!("malformed_output_{name}_{parser_type}"),
                        rendered
                    );
                });
            })
    }

    #[test]
    fn test_render() {
        let renderer = DiffRenderer::new();
        let testcase = TestCase {
            title: "the title".into(),
            shell_expression: "the command".into(),
            expectations: vec![
                test_expectation!("expected line 1"),
                test_expectation!("expected line 2"),
                test_expectation!("expected line 3"),
            ],
            exit_code: None,
            line_number: 234,
            ..Default::default()
        };

        let tests = vec![
            (
                "missing",
                Diff::new(vec![DiffLine::UnmatchedExpectation {
                    index: 1,
                    expectation: testcase.expectations[1].clone(),
                }]),
            ),
            (
                "unexpected",
                Diff::new(vec![DiffLine::UnexpectedLines {
                    lines: vec![(2, "something else".as_bytes().to_vec())],
                }]),
            ),
            (
                "mismatch",
                Diff::new(vec![
                    DiffLine::UnmatchedExpectation {
                        index: 1,
                        expectation: testcase.expectations[1].clone(),
                    },
                    DiffLine::UnmatchedExpectation {
                        index: 2,
                        expectation: testcase.expectations[2].clone(),
                    },
                    DiffLine::UnexpectedLines {
                        lines: vec![
                            (2, "something line 1".as_bytes().to_vec()),
                            (3, "something line 2".as_bytes().to_vec()),
                        ],
                    },
                ]),
            ),
        ];

        [ParserType::Cram, ParserType::Markdown]
            .iter()
            .for_each(|parser_type| {
                tests.iter().for_each(|(name, diff)| {
                    let mut testcase1 = testcase.clone();
                    testcase1.line_number = 10;
                    let mut testcase2 = testcase.clone();
                    testcase2.line_number = 20;
                    let outcomes = vec![
                        Outcome {
                            output: (
                                "expected line 1\nexpected line FAIL\nexpected line 3\n",
                                "the stderr",
                            )
                                .into(),
                            testcase: testcase2,
                            location: Some("location2".into()),
                            result: Err(TestCaseError::MalformedOutput(diff.to_owned())),
                            format: *parser_type,
                            escaping: Escaper::default(),
                        },
                        Outcome {
                            output: (
                                "expected line 1\nexpected line FAIL\nexpected line 3\n",
                                "the stderr",
                            )
                                .into(),
                            testcase: testcase1,
                            location: Some("location1".into()),
                            result: Err(TestCaseError::MalformedOutput(diff.to_owned())),
                            format: *parser_type,
                            escaping: Escaper::default(),
                        },
                    ];
                    let rendered = renderer
                        .render(&outcomes.iter().collect::<Vec<_>>())
                        .expect("render succeeds");
                    insta::assert_snapshot!(format!("render_{name}_{parser_type}"), rendered);
                });
            })
    }
}
