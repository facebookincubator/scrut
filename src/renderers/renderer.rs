/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use anyhow::Result;

use crate::diff::Diff;
use crate::outcome::Outcome;
use crate::testcase::TestCaseError;

/// Renderer translate errors from validating [`crate::testcase::TestCase`]s into
/// text that allows humans to understand how the actual output is different
/// from the expected output
pub trait Renderer {
    fn render(&self, outcomes: &[&Outcome]) -> Result<String>;
}

/// This seems like a good idea atm? Assuming there will be more renderers Maybe not ..
pub(super) trait ErrorRenderer: Renderer {
    fn render_error(&self, err: &TestCaseError, outcome: &Outcome) -> Result<String> {
        match err {
            TestCaseError::MalformedOutput(diff) => self.render_malformed_output(outcome, diff),
            TestCaseError::InvalidExitCode { actual, expected } => {
                self.render_invalid_exit_code(outcome, *actual, *expected)
            }
            TestCaseError::InternalError(err) => self.render_delegated_error(outcome, err),
            TestCaseError::Skipped => self.render_skipped(outcome),
        }
    }

    fn render_invalid_exit_code(
        &self,
        outcome: &Outcome,
        actual: i32,
        expected: i32,
    ) -> Result<String>;

    fn render_delegated_error(&self, outcome: &Outcome, err: &anyhow::Error) -> Result<String>;

    fn render_malformed_output(&self, outcome: &Outcome, diff: &Diff) -> Result<String>;

    fn render_skipped(&self, outcome: &Outcome) -> Result<String>;
}
