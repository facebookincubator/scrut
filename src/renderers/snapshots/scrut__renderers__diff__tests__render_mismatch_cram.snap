---
source: src/renderers/diff.rs
assertion_line: 546
expression: rendered
---
--- location1
+++ location1.new
@@ -12,2 +13,2 @@ malformed output: the title
-  expected line 2
-  expected line 3
+  something line 1
+  something line 2

--- location2
+++ location2.new
@@ -22,2 +23,2 @@ malformed output: the title
-  expected line 2
-  expected line 3
+  something line 1
+  something line 2

